FeedStoreChallenge/CoreDataFeedStore.swift

```diff
@@ -0,0 +1,107 @@
+//
+//  CoreDataFeedStore.swift
+//  FeedStoreChallenge
+//
+//  Created by Mohammed Al Waili on 01/03/2021.
+//  Copyright © 2021 Essential Developer. All rights reserved.
+//
+
+import Foundation
+import CoreData
+
+final public class CoreDataFeedStore: FeedStore {
+	
+	private let persistentContainer: NSPersistentContainer
+	private let context: NSManagedObjectContext
+	
+	public init(storeURL: URL, bundle: Bundle = .main) throws {
+		self.persistentContainer = try NSPersistentContainer.load(modelName: "LocalFeedImageModel", url: storeURL, in: bundle)
+		self.context = self.persistentContainer.newBackgroundContext()
+	}
+	
+	public func deleteCachedFeed(completion: @escaping DeletionCompletion) {
+		perform { context in
+			do {
+				if let cache = try CacheEntity.fetch(in: context) {
+					context.delete(cache)
+				}
+				try context.save()
+				completion(nil)
+			} catch {
+				completion(error)
+			}
+		}
+	}
+	
+	public func insert(_ feed: [LocalFeedImage], timestamp: Date, completion: @escaping InsertionCompletion) {
+		perform { context in
+			do {
+				let newCache = try CacheEntity.newInstance(in: context)
+				newCache.timestamp = timestamp
+				newCache.feed = NSOrderedSet(array: LocalFeedImageEntity.entities(from: feed, in: context))
+				try context.save()
+				completion(nil)
+			} catch {
+				completion(error)
+			}
+		}
+		
+	}
+	
+	public func retrieve(completion: @escaping RetrievalCompletion) {
+		perform { context in
+			do {
+				if let cache = try CacheEntity.fetch(in: context) {
+					completion(.found(feed: CacheEntity.localFeed(from: cache.feed), timestamp: cache.timestamp))
+				} else {
```

Comment:
If you follow my previous recommendation about the computed var, here's how simpler this expression can be:

```suggestion
				if let cache = try CacheEntity.fetch(in: context) {
					completion(.found(feed: cache.localFeed, timestamp: cache.timestamp))
				} else {
```